### ISSUE TITLE
Updating of repositories crashes

### ISSUE THREAD

**Issue by bgruening** [galaxyproject/galaxy#888]
Updating of repositories can lead to the following crash. Tested for example with `package_bzlib_1_0` from the TTS.  Reinstalling the repository works fine.

```
Traceback (most recent call last):
  File "lib/galaxy/web/framework/middleware/request_id.py", line 14, in __call__
    return self.app( environ, start_response )
  File "lib/galaxy/web/framework/middleware/xforwardedhost.py", line 22, in __call__
    return self.app( environ, start_response )
  File "lib/galaxy/web/framework/middleware/translogger.py", line 70, in __call__
    return self.application(environ, replacement_start_response)
  File "lib/galaxy/web/framework/middleware/error.py", line 163, in __call__
    response = self.exception_handler(exc_info, environ)
  File "lib/galaxy/web/framework/middleware/error.py", line 178, in exception_handler
    get_vars = wsgilib.parse_querystring(environ)
  File "/galaxy-central/eggs/Paste-1.7.5.1-py2.7.egg/paste/request.py", line 100, in parse_querystring
    strict_parsing=False)
  File "/usr/lib/python2.7/cgi.py", line 190, in parse_qsl
    PendingDeprecationWarning, 2)
  File "lib/galaxy/web/framework/middleware/error.py", line 151, in __call__
    app_iter = self.application(environ, sr_checker)
  File "/galaxy-central/eggs/Paste-1.7.5.1-py2.7.egg/paste/recursive.py", line 84, in __call__
    return self.application(environ, start_response)
  File "/galaxy-central/eggs/Paste-1.7.5.1-py2.7.egg/paste/httpexceptions.py", line 633, in __call__
    return self.application(environ, start_response)
  File "lib/galaxy/web/framework/base.py", line 133, in __call__
    return self.handle_request( environ, start_response )
  File "lib/galaxy/web/framework/base.py", line 191, in handle_request
    body = method( trans, **kwargs )
  File "lib/galaxy/web/framework/decorators.py", line 92, in decorator
    return func( self, trans, *args, **kwargs )
  File "lib/galaxy/webapps/galaxy/controllers/admin_toolshed.py", line 1872, in update_to_changeset_revision
    hg_util.update_repository( repo, latest_ctx_rev )
  File "lib/tool_shed/util/hg_util.py", line 461, in update_repository
    commands.update( get_configured_ui(), repo, rev=ctx_rev )
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/commands.py", line 6255, in update
    ret = hg.update(repo, rev)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/hg.py", line 499, in update
    stats = updaterepo(repo, node, False)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/hg.py", line 495, in updaterepo
    labels=['working copy', 'destination'])
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/merge.py", line 1122, in update
    stats = applyupdates(repo, actions, wc, p2, overwrite, labels=labels)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/merge.py", line 687, in applyupdates
    r = ms.resolve(f, wctx, labels=labels)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/merge.py", line 291, in resolve
    labels=labels)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/filemerge.py", line 424, in filemerge
    (a, b, c, back), labels=labels)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/filemerge.py", line 279, in _xmerge
    r = _premerge(repo, toolconf, files, labels=labels)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/filemerge.py", line 201, in _premerge
    r = simplemerge.simplemerge(ui, a, b, c, quiet=True, label=labels)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/simplemerge.py", line 371, in simplemerge
    out = opener(os.path.basename(local), "w", atomictemp=True)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/scmutil.py", line 337, in __call__
    self.audit(path)
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/pathutil.py", line 46, in __call__
    or _lowerclean(parts[0]) in ('.hg', '.hg.', '')
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/pathutil.py", line 8, in _lowerclean
    return encoding.hfsignoreclean(s.lower())
  File "/galaxy-central/eggs/mercurial-3.2.4-py2.7-linux-x86_64-ucs4.egg/mercurial/encoding.py", line 28, in hfsignoreclean
    if "\xe2" in s or "\xef" in s:
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 0: ordinal not in range(128)
```

**Comment by chambm** [comment]
I also see this quite frequently. First in 15.10-devish, and still in 16.04 release.

**Comment by Public-Health-Bioinformatics** [comment]
I'm getting this too on a tool I'm developing.  First install from toolshed works fine, but an update from a new (minor edit on) version from toolshed fails with above error. However I don't know the version of Galaxy that's installed - its not a git or hg repo.  I can find out from original installer if needed?  Can this be avoided with proper python "# -_\- coding: utf-8 -_-" or is it something else?

**Comment by mvdbeek** [comment]
we're now using the external mercurial script, if you see this again feel free to re-open.

**Comment by ggaffield** [comment]
I experienced this in the 18.01 release, which is pulled into the current release of bgruening/galaxy-rna-workbench.  To reproduce: Admin/Manage Tools --> "Update tool shed status" (when it returns, pick a tool that shows "Updates are available..." (in my case bowtie2), then select action "Get updates".  You'll get a blank screen and the "UnicodeDecodeError" in the uwsgi log.
  
I resolved this locally by editing mercurial/encoding.py to test if variable "s" was already encoded or not.  Maybe not the best solution, but it got me going again.
```
    try:
        s.decode('utf-8')
        if "\xe2" in s or "\xef" in s:
            for c in _ignore:
                s = s.replace(c, '')
    except:
        pass
    return s
```
